pkg/vpcmodel/subnetsIPBlocks.go

```diff
@@ -0,0 +1,252 @@
+	}
+	address := subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex].FirstIPAddress()
+	return address, subnetsBlocks.removeAddressFromFree(address, subnetCRN, blockIndex)
+}
+func (subnetsBlocks SubnetsIPBlocks) SubnetBlocks(subnetCRN string) []*ipblock.IPBlock {
+	return subnetsBlocks[subnetCRN].splitByFiltersBlocks
+}
+func (subnetsBlocks SubnetsIPBlocks) IsFullyReservedBlock(subnetCRN string, blockIndex int) bool {
+	return subnetsBlocks[subnetCRN].fullyReservedBlocks[blockIndex]
+}
+
+func (subnetsBlocks SubnetsIPBlocks) removeAddressFromFree(address, subnetCRN string, blockIndex int) error {
+	addressBlock, err := ipblock.FromIPAddress(address)
+	if err != nil {
+		return err
+	}
+	subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex] =
+		subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex].Subtract(addressBlock)
+	return nil
+}
+
+// filtersBlocks is a map from VPC UID to a list of IPBlocks,
+// which holds the list of disjoint IPBlocks computed from all referenced CIDRs in that VPC's filters rules
+type filtersBlocks map[string][]*ipblock.IPBlock
+
+// ///////////////////////////////////////////
+// getFiltersBlocks() create a slice of disjoint blocks for each vpc, split according to the rules blocks, and their sum is allCidr:
+func getFiltersBlocks(filtersCidrs []map[string][]*string) (blocks filtersBlocks, err error) {
+	// sorting by vpc:
+	vpcCidrs := getVpcsCidrs(filtersCidrs)
+	// disjointing:
+	return disjointVpcCidrs(vpcCidrs)
+}
+
+// getVpcsCidrs() sort the cidr of all filters for each vpc separately, adding ipblock.CidrAll for every vpc
+func getVpcsCidrs(filtersCidrs []map[string][]*string) map[string][]string {
+	vpcCidrs := map[string][]string{}
+	for _, filterCidrsOrAddresses := range filtersCidrs {
+		for vpc, vpcCidrsOrAddresses := range filterCidrsOrAddresses {
+			if _, ok := vpcCidrs[vpc]; !ok {
+				vpcCidrs[vpc] = []string{ipblock.CidrAll}
+			}
+			for _, cidrOrAddress := range vpcCidrsOrAddresses {
+				if cidrOrAddress != nil {
+					vpcCidrs[vpc] = append(vpcCidrs[vpc], *cidrOrAddress)
+				}
+			}
+		}
+	}
+	return vpcCidrs
+}
+
+// disjointVpcCidrs() disjoint the cidr for each vpc.
+func disjointVpcCidrs(cidr map[string][]string) (blocks filtersBlocks, err error) {
+	blocks = filtersBlocks{}
+	for vpc, vpcCidr := range cidr {
+		blocks[vpc], err = disjointCidrs(vpcCidr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return blocks, nil
+}
+
+// disjointCidrs() get a slice of cidrs/addresses can create a slice of disjoint blocks.
+// the algorithm:
+//  1. remove duplicate cidrs
+//  2. convert the cidr to a slice of blocks
+//  3. sort the blocks by size - small to big
+//  4. iterate over the blocks: for each block, create a disjoint block by subtracting from the block the bormer blocks
+//
+// todo - move this function to modules?
+func disjointCidrs(cidrs []string) ([]*ipblock.IPBlock, error) {
+	compactCidrs := slices.Compact(cidrs)
```

Comment:
right
this compact its "nice to have", to save time, and the algorithm will work without it.
I can add another compact, on the block level, but I dont think it will make it faster